bpf/gotracer/go_sql.c

```diff
@@ -25,11 +25,99 @@
 #include <common/ringbuf.h>
 
 #include <gotracer/go_common.h>
+#include <gotracer/go_str.h>
+
+// Extracts MySQL server hostname by following the pointer chain:
+// driverConn -> ci (interface) -> mysqlConn -> cfg (*Config) -> Addr (string)
+// Returns the Addr string which has format "host:port"
+static __always_inline int
+read_mysql_hostname_from_driverconn(u64 driver_conn_ptr, char *hostname, int max_len) {
+    if (driver_conn_ptr == 0) {
+        bpf_dbg_printk("can't read mysql hostname: driver_conn_ptr is null");
+        return 0;
+    }
+
+    off_table_t *ot = get_offsets_table();
+    if (!ot) {
+        bpf_dbg_printk("can't read mysql hostname: offset table not available");
+        return 0;
+    }
+
+    // driverConn.ci is a Go interface [type_ptr (8 bytes), data_ptr (8 bytes)]
+    // Read the data pointer (at ci_offset + 8) to get the actual mysqlConn instance
+    void *mysql_conn_ptr = 0;
+    int res = bpf_probe_read(
+        &mysql_conn_ptr,
+        sizeof(mysql_conn_ptr),
+        (void *)(driver_conn_ptr + go_offset_of(ot, (go_offset){.v = _driverconn_ci_pos}) + 8));
+
+    if (res || !mysql_conn_ptr) {
+        bpf_dbg_printk("can't read database/sql.driverConn.ci");
+        return 0;
+    }
+
+    // Dereference mysqlConn.cfg to get pointer to Config struct
+    void *cfg_ptr = 0;
+    res = bpf_probe_read(
+        &cfg_ptr,
+        sizeof(cfg_ptr),
+        (void *)((u64)mysql_conn_ptr + go_offset_of(ot, (go_offset){.v = _mysql_conn_cfg_pos})));
+
+    if (res || !cfg_ptr) {
+        bpf_dbg_printk("can't read mysql.mysqlConn.cfg");
+        return 0;
+    }
+
+    // Read Config.Addr string field
+    if (!read_go_str("mysql hostname",
+                     cfg_ptr,
+                     go_offset_of(ot, (go_offset){.v = _mysql_config_addr_pos}),
+                     hostname,
+                     max_len)) {
+        bpf_dbg_printk("can't read mysql.Config.Addr");
+        return 0;
+    }
 
-static __always_inline void set_sql_info(void *goroutine_addr, void *sql_param, void *query_len) {
+    return 1;
+}
+
+static __always_inline void extract_sql_hostname(sql_request_trace_t *trace, u64 driver_conn_ptr) {
+    trace->hostname[0] = '\0';
+
+    if (driver_conn_ptr == 0) {
+        bpf_dbg_printk("sql hostname extraction skipped: driver_conn_ptr is null");
+        return;
+    }
+
+    off_table_t *ot = get_offsets_table();
+    if (!ot) {
+        bpf_dbg_printk("sql hostname extraction failed: offset table not available");
+        return;
+    }
+
+    void *driver_impl_ptr = 0;
```

Comment:
this code is duplicate from above, perhaps it can be extracted as a helper function